packages/ember-glimmer/tests/integration/syntax/with-test.js

```diff
@@ -30,4 +38,222 @@ moduleFor('Syntax test: {{#with}}', class extends SharedConditionalsTest {
+  }
+
+  ['@test re-using the same variable with different #with blocks does not override each other']() {
+    this.render(`Admin: {{#with admin as |person|}}{{person.name}}{{/with}} User: {{#with user as |person|}}{{person.name}}{{/with}}`, {
+      admin: { name: 'Tom Dale' },
+      user: { name: 'Yehuda Katz' }
+    });
+
+    this.assertText('Admin: Tom Dale User: Yehuda Katz');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Admin: Tom Dale User: Yehuda Katz');
+
+    this.runTask(() => {
+      set(this.context, 'admin.name', 'Erik Bryn');
+      set(this.context, 'user.name', 'Chad Hietala');
+    });
+
+    this.assertText('Admin: Erik Bryn User: Chad Hietala');
+
+    this.runTask(() => {
+      set(this.context, 'admin', { name: 'Tom Dale' });
+      set(this.context, 'user', { name: 'Yehuda Katz' });
+    });
+
+    this.assertText('Admin: Tom Dale User: Yehuda Katz');
+  }
+
+  ['@test the scoped variable is not available outside the {{with}} block.']() {
+    this.render(`{{#with first as |ring|}}{{ring}}-{{#with fifth as |ring|}}{{ring}}-{{#with ninth as |ring|}}{{ring}}-{{/with}}{{ring}}-{{/with}}{{ring}}{{/with}}`, {
+      first: 'Limbo',
+      fifth: 'Wrath',
+      ninth: 'Treachery'
+    });
+
+    this.assertText('Limbo-Wrath-Treachery-Wrath-Limbo');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Limbo-Wrath-Treachery-Wrath-Limbo');
+
+    this.runTask(() => {
+      set(this.context, 'first', 'I');
+      set(this.context, 'fifth', 'D');
+      set(this.context, 'ninth', 'K');
+    });
+
+    this.assertText('I-D-K-D-I');
+
+    this.runTask(() => {
+      set(this.context, 'first', 'Limbo');
+      set(this.context, 'fifth', 'Wrath');
+      set(this.context, 'ninth', 'Treachery');
+    });
+
+    this.assertText('Limbo-Wrath-Treachery-Wrath-Limbo');
+  }
+
+  ['@test it should support #with name as |foo|, then #with foo as |bar|']() {
+    this.render(`{{#with name as |foo|}}{{#with foo as |bar|}}{{bar}}{{/with}}{{/with}}`, {
+      name: 'caterpillar'
+    });
+
+    this.assertText('caterpillar');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('caterpillar');
+
+    this.runTask(() => set(this.context, 'name', 'butterfly'));
+
+    this.assertText('butterfly');
+
+    this.runTask(() => set(this.context, 'name', 'caterpillar'));
+
+    this.assertText('caterpillar');
+  }
+
+  ['@test nested {{with}} blocks shadow the outer scoped variable properly.']() {
+    this.render(`{{#with first as |ring|}}{{ring}}-{{#with fifth as |ring|}}{{ring}}-{{#with ninth as |ring|}}{{ring}}-{{/with}}{{ring}}-{{/with}}{{ring}}{{/with}}`, {
+      first: 'Limbo',
+      fifth: 'Wrath',
+      ninth: 'Treachery'
+    });
+
+    this.assertText('Limbo-Wrath-Treachery-Wrath-Limbo');
+
+    this.runTask(() => this.rerender());
+
+    this.runTask(() => {
+      set(this.context, 'first', 'I');
+      set(this.context, 'fifth', 'D');
+      set(this.context, 'ninth', 'K');
+    });
```

Comment:
Can you either 1) break this up into two `update` cycles or 2) delete one of these?

There is a failure mode where when updating, it clobbers either an upstream/downstream slot, but that might not happen when you update *all* of the slots at the same time. So I would like to add a step (or two) where we are updating **some** of the values, but not all.